Make hasSpec tolerate more ghci settings.

fisx opened the pull request:
If you have something like `:set +m +t` in your global .ghci file, hasSpec will get confused by the more verbose output.  This change matches against all lines of the output until it succeeds.

If you think this change is too hacky, I'd be interested if you have another idea.

fisx commented:
I think there is at least one more place where the same issue occurs.  I'll look into this a little more.

fisx commented:
Ok, back to you.  Did I miss anything else?

(Also: this is awesome!  Thanks!  :-)

sol commented:
I think the right way to address this is to ignore `~/.ghci` and only source project specific `.ghci` files.  Sadly, from what I tried `-ignore-dot-ghci` and `-ghci-script` can not be used in combination, e.g.
```
ghci test/Spec.hs -ignore-dot-ghci  -ghci-script .ghci
```
does not source `.ghci`.

(https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/ghci-dot-files.html)

May issue with the global `.ghci` is that it may contain all kinds of things that may lead to issues.

@fisx any input on this?

(All that said, we may still try to make things more robust, and this PR looks like a good start.)

fisx commented:
Ok, I will look into this some more.  I agree that it would be nice to have a more general solution.

fisx commented:
`-ignore-dot-ghci` is a good work-around for this problem, as you can always do anything that would otherwise happen in a local .ghci on the command line.  The only better solution I can think of that is not insane is changing the baviour of ghci.

I would still be happy to see my changes merged: I don't think they introduce any new problems, and they make the work-around unnecessary at least in my case, and probably in many others.

Up to you, obviously.

sol commented on src/Session.hs:
```diff
@@ -92,10 +92,10 @@ isSuccess :: Maybe Summary -> Bool
-parseSummary r = case reverse $ lines r of
-  x : _ | Just summary <- readMaybe (dropAnsiEscapeSequences x) -> Just summary
-  _ -> Nothing
+parseSummary = findJust . (map $ readMaybe . dropAnsiEscapeSequences) . take 3 . reverse . lines
```
What about skipping the `take 3`.  Would it make it even more robust?  Is there anything that can go wrong?

fisx replied:
If any of the lines can be parsed as a summary even though they are not, the test result will be wrong.  In order to minimize that from happening, I limited the window that is examined.

I admit that this is an esoteric problem, so if you poke me once more I'll drop the `take 3` without further discussion.

sol replied:
My reasoning is that at this point we already know there is a spec, and running a spec will always yield a summary. So I would hope that taking the first thing, starting from the back of the list, that looks like a summary, is fine.

sol commented on src/Session.hs:
```diff
@@ -92,10 +92,10 @@ isSuccess :: Maybe Summary -> Bool
 isSuccess = not . isFailure
 
 parseSummary :: String -> Maybe Summary
-parseSummary r = case reverse $ lines r of
```
I think at this point we want tests for `parseSummary`.  It document's your change and makes sure that it will not get lost with a future refactoring.

sol commented on src/Session.hs:
```diff
@@ -70,9 +71,8 @@ hspecCommand = "Test.Hspec.Runner.hspecResult spec"
-    [ys] -> return $ (hspecCommand ++ " :: IO ") `isPrefixOf` ys && "Summary" `isSuffixOf` ys
-    _ -> return False
+  let match line = (hspecCommand ++ " :: IO ") `isPrefixOf` line && "Summary" `isSuffixOf` line
+  return . any match . lines $ xs
```
Again, it would be nice to test this change.  Probably it makes sense to split `hasSpec` into a pure and an `IO` part and focus on testing the pure part.

sol commented:
@fisx ok, let's do that, let's get this on `master`.  I reviewed your changes and added some comments.  Can you rebase on current `master` and address them?

soenkehahn commented:
Btw, I have `:set -fdefer-typed-holes` in my `~/.ghci` and think it's very convenient that `sensei` picks it up. Ignoring that file seems like a bad idea to me for that reason.

fisx commented:
@sol: I think I addressed all your comments.  I like the code much better now (even though I would probably go with fewer tests. :-).

Anything else?

sol commented:
@soenkehahn good point.

fisx commented:
> My reasoning is that at this point we already know there is a spec, and running a spec will always yield a summary. So I would hope that taking the first thing, starting from the back of the list, that looks like a summary, is fine.

Right.  Should have thought of that myself.  Thanks!

sol commented:
@fisx yes, less tests are fine here.  In general I think it's completely valid to not test exhaustively, but only write the tests that are necessary to drive the implementation.  This is how I would do it: https://github.com/hspec/sensei/commit/ccb093661ad71d698c0c246a85b77b71d04c2c88

fisx commented:
Yes, your's looks better.  You could do `Summary 2 0` -> `summary` in the last line, but I guess it works as is.

Thanks for the thorough review!
